go/tools/releasing/cmd/prerelease.go

```diff
@@ -0,0 +1,304 @@
+	cmd := exec.Command("git", "diff", "--exit-code")
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("working tree is not clean, can't proceed with the release process:\n\n%v",
+			string(output),
+		)
+	}
+
+	return nil
+}
+
+func (p prerelease) createPrereleaseBranch(fromExistingBranch string) error {
+	branchNameElements := []string{"pre_release", p.modSetName, p.newVersion}
+	branchName := strings.Join(branchNameElements, "_")
+	fmt.Printf("git checkout -b %v %v\n", branchName, fromExistingBranch)
+	cmd := exec.Command("git", "checkout", "-b", branchName, fromExistingBranch)
+
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("could not create new branch %v: %v (%v)", branchName, string(output), err)
+	}
+
+	return nil
+}
+
+// TODO: updateVersionGo may be implemented to update any hard-coded values within version.go files as needed.
+func (p prerelease) updateVersionGo() error {
+	return nil
+}
+
+// runMakeLint runs 'make lint' to automatically update go.sum files.
+func (p prerelease) runMakeLint() error {
+	fmt.Println("Updating go.sum with 'make lint'...")
+
+	cmd := exec.Command("make", "lint")
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("'make lint' failed: %v (%v)", string(output), err)
+	}
+
+	return nil
+}
+
+func (p prerelease) commitChanges(skipMake bool) error {
+	commitMessage := "Prepare for releasing " + p.newVersion
+
+	// add changes to git
+	cmd := exec.Command("git", "add", ".")
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("'git add .' failed: %v (%v)", string(output), err)
+	}
+
+	// make ci
+	if skipMake {
+		fmt.Println("Skipping 'make ci'...")
+	} else {
+		fmt.Println("Running 'make ci'...")
+		cmd = exec.Command("make", "ci")
+		if output, err := cmd.CombinedOutput(); err != nil {
+			return fmt.Errorf("'make ci' failed: %v (%v)", string(output), err)
+		}
+	}
+
+	// commit changes to git
+	fmt.Printf("Commit changes to git with message '%v'...\n", commitMessage)
+	cmd = exec.Command("git", "commit", "-m", commitMessage)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("git commit failed: %v (%v)", string(output), err)
+	}
+
+	cmd = exec.Command("git", "log", `--pretty=format:"%h"`, "-1")
+	output, err := cmd.Output()
+	if err != nil {
+		fmt.Println("WARNING: could not automatically get last commit hash.")
+	}
+
+	fmt.Println("Commit successful. Hash of commit:")
+	os.Stdout.Write(output)
+
+	return nil
+}
+
+// updateGoModVersions reads the fromFile (a go.mod file), replaces versions
+// for all specified modules in newModPaths, and writes the new go.mod to the toFile file.
+func (p prerelease) updateGoModVersions(modFilePath tools.ModuleFilePath) error {
+	newGoModFile, err := ioutil.ReadFile(string(modFilePath))
+	if err != nil {
+		panic(err)
+	}
+
+	for _, modPath := range p.modPaths {
+		oldVersionRegex := filePathToRegex(string(modPath)) + ` v[0-9]*\.[0-9]*\.[0-9]`
```

Comment:
```suggestion
		oldVersionRegex := regexp.QuoteMeta(string(modPath)) + `\s+v\d+\.\d+\.\d+`
```

https://pkg.go.dev/regexp#QuoteMeta will safely quote all meta characters.

The regexp part covering the version probably needs some work, I don't think this will properly handle valid versions such as `v1.0.0-RC1`.

Reply:
Thanks for the QuoteMeta suggestion! I was sure there was something like it but didn't find it in my search 👍 

How about this for a semver Regex? 😆 You can see it tested [here](https://regexr.com/6145k), inspired by [this](https://regexr.com/39s32) and [this](https://ihateregex.io/expr/semver/)
\s+v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`